From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ryan Volz <ryan.volz@gmail.com>
Date: Fri, 14 Mar 2025 10:18:10 -0400
Subject: [PATCH] Wait until all transfers are cancelled

This fixes a crash when stopping and restarting a data stream on Windows
---
 src/librtlsdr.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/src/librtlsdr.c b/src/librtlsdr.c
index a8aae48..b6fd785 100644
--- a/src/librtlsdr.c
+++ b/src/librtlsdr.c
@@ -2180,8 +2180,16 @@ int rtlsdr_read_async(rtlsdr_dev_t *dev, rtlsdr_read_async_cb_t cb, void *ctx,
 #endif
 					libusb_handle_events_timeout_completed(dev->ctx,
 									       &zerotv, NULL);
-					if (r < 0)
+					if (r == LIBUSB_ERROR_NOT_FOUND) {
+						/* transfer not in progress, already complete, or
+						 * already cancelled, but we're here because the
+						 * handler has not yet been called, so keep waiting */
+						libusb_handle_events_timeout_completed(dev->ctx,
+															   &tv,
+															   NULL);
+					} else if (r < 0) {
 						continue;
+					}
 
 					next_status = RTLSDR_CANCELING;
 				}
-- 
2.47.1

